src/levanter/schedule.py

```diff
@@ -0,0 +1,122 @@
+import typing
+from dataclasses import dataclass
+from typing import Sequence, TypeVar
+
+
+T = TypeVar("T")
+
+
+@dataclass
+class ScheduleStep(typing.Generic[T]):
+    until: int
+    value: T
+
+
+IntSchedule = list[ScheduleStep[int]]
+
+BIG_INT = 2**63 - 1
+
+
+def value_at_step(schedule_or_t: Sequence[ScheduleStep[T]] | T, step: int) -> T:
+    """
+    Given a schedule or a single value, return the value at the given step.
+
+    """
+    if not isinstance(schedule_or_t, Sequence) or (schedule_or_t and not isinstance(schedule_or_t[0], ScheduleStep)):
+        return schedule_or_t  # type: ignore
+
+    for i, step_ in enumerate(schedule_or_t):
+        if step < step_.until or step_.until == -1:
+            return step_.value
+
+    return schedule_or_t[-1].value
+
+
+def validate_schedule_sorted(schedule: Sequence[ScheduleStep[T]]):
+    for i in range(1, len(schedule)):
+        # the last step can be -1 to indicate that the value should be held indefinitely
+        if schedule[i].until < schedule[i - 1].until:
+            if i == len(schedule) - 1 and schedule[i].until == -1:
+                continue
+            raise ValueError(f"Schedule is not sorted at index {i}")
+
+
+@dataclass
+class BatchSegment:
+    start: int  # The training step at which this batch size starts.
+    until: int | float  # The training step at which this batch size stops. If -1 or inf, the segment is open-ended.
```

Comment:
Maybe support just -1 or inf for simplicity sake.

Reply:
Doubly so with the BIG_INT mixed in. Some calls support until being natively -1, while the below overrides -1 to BIG_INT and then assumes that in the logic.

Reply:
oh yeah i was gonna support inf but it got annoying so I started to remove it but didn't finish